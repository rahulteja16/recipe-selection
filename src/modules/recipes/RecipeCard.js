import React from 'react';

import Box from '../../components/Box';
import Button from '../../components/Button';
import Flex from '../../components/Flex';
import IconMinusCircle from '../../icons/IconMinusCircle';
import IconPlusCircle from '../../icons/IconPlusCircle';
import PropTypes from 'prop-types';
import Text from '../../components/Text';
import css from '@styled-system/css';
import { parseRawPrice } from './price';
import styled from 'styled-components';

const RecipeCard = ({
  extraCharge,
  headline,
  id,
  image,
  isAddEnabled,
  name,
  selected,
  yields,
  handleAddRecipe,
  handleRemoveRecipe,
  maxRecipesSelected,
  selectedRecipes,
  isExtra,
  isMaxLimitReached
}) =>  {
  let isDisabled = false;
  if(isMaxLimitReached) {
    isDisabled = true;
  } else {
    if(isAddEnabled) {
      isDisabled = selectedRecipes <= maxRecipesSelected
    } else {
      isDisabled = true;
    }
  }

  return   <Box
  data-testid="recipe-card"
  borderWidth={selected ? 'md' : null}
  borderStyle={selected ? 'solid' : null}
  borderColor={selected ? 'primary_600' : null}
  m={selected ? '-2px' : null}
  borderRadius="md"
  boxShadow="lg">
  <Box borderRadius="2px 2px 0px 0px" paddingBottom="56.25%" overflow="hidden" height="0">
    <img src={image} alt={name} width="100%" />
  </Box>

  <Box p="xs" height="120px">
    <Text  data-testid="recipe-card-name" fontWeight="bold" fontFamily="primary" fontSize="md">
      {name}
    </Text>
    <Text data-testid="recipe-card-headline" fontWeight="regular" fontFamily="secondary" fontSize="sm">
      {headline}
    </Text>
  </Box>
  {selected ? (
    <SelectedRecipeFooter
      recipeId={id}
      yields={yields}
      selected={selected}
      isDisabled={isDisabled}
      handleAddRecipe={handleAddRecipe}
      handleRemoveRecipe={handleRemoveRecipe}
    />
  ) : (
    <UnselectedRecipeFooter
      recipeId={id}
      price={extraCharge && extraCharge}
      handleAddRecipe={handleAddRecipe}
      isExtra={isExtra}
      isDisabled={isDisabled}
    />
  )}
</Box>
};

RecipeCard.propTypes = {
  extraCharge: PropTypes.number,
  handleAddRecipe: PropTypes.func,
  handleRemoveRecipe: PropTypes.func,
  headline: PropTypes.string,
  id: PropTypes.string,
  image: PropTypes.string,
  maxRecipesSelected: PropTypes.bool,
  minRecipesSelected: PropTypes.bool,
  name: PropTypes.string,
  selected: PropTypes.number,
  selectionLimit: PropTypes.number,
  yields: PropTypes.number,
};

export const UnselectedRecipeFooter = ({
  isExtra,
  price,
  recipeId,
  handleAddRecipe,
  isDisabled
}) => (
  <Flex p="xs" data-testid="recipe-unselected">
    <Box flex="50%" alignSelf="center">
      {price ? <Text data-testid="recipe-special-price" color="primary_600">+{parseRawPrice(price)}</Text> : null}
    </Box>
    <Box flex="50%">
      <Button
        onClick={() => handleAddRecipe(recipeId, 1)}
        variant="secondary"
        width="100%"
        p="0"
        disabled={isDisabled}>
        {isExtra  ? 'Add extra meal' : 'Add'}
      </Button>
    </Box>
  </Flex>
);

UnselectedRecipeFooter.propTypes = {
  isExtra: PropTypes.bool,
  price: PropTypes.number,
  recipeId: PropTypes.string,
  isDisabled: PropTypes.bool,
  handleAddRecipe: PropTypes.func,
};

export const SelectedRecipeFooter = ({
  recipeId,
  selected,
  yields,
  handleAddRecipe,
  handleRemoveRecipe,
  isDisabled
}) => {

return (
  <Flex backgroundColor="primary_600" justifyContent="space-between" alignItems="center" data-testid="recipe-selected">
    <SelectionButton onClick={() => handleRemoveRecipe(recipeId, selected - 1)} title="Decrease quantity" aria-label='decrease-quantity'>
      <IconMinusCircle />
    </SelectionButton>
    <Box color="white">
      <Text textAlign="center" fontWeight="bold" fontFamily="secondary" fontSize="md" data-testid="recipe-selected-desc">
        {selected} in your box
      </Text>
      <Text textAlign="center" fontFamily="secondary" fontSize="sm" data-testid="recipe-selected-servings">
        ({selected * yields} servings)
      </Text>
    </Box>
    <SelectionButton onClick={() => handleAddRecipe(recipeId, selected + 1)} title="Increase quantity" disabled={isDisabled} aria-label='increase-quantity'>
      <IconPlusCircle />
    </SelectionButton>
  </Flex>
)
}

SelectedRecipeFooter.propTypes = {
  recipeId: PropTypes.string,
  selected: PropTypes.number,
  yields: PropTypes.number,
  handleAddRecipe: PropTypes.func,
  handleRemoveRecipe: PropTypes.func,
  isDisabled: PropTypes.bool
};

const SelectionButton = styled.button`
  ${css({
    outline: 'none',
    color: 'white',
    padding: 'sm',
    cursor: 'pointer',
    backgroundColor: 'primary_600',
    border: 'none',
    ':hover:enabled': {
      backgroundColor: 'primary_700',
    },
    ':active:enabled': {
      backgroundColor: 'primary_800',
    },
  })}
  &:disabled {
    cursor: default;
    opacity: 0.5;
  }
`;

export default RecipeCard;
