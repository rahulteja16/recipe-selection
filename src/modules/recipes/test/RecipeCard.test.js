import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import RecipeCard, { UnselectedRecipeFooter,  SelectedRecipeFooter} from '../RecipeCard';

const recipeCardProps = {
  extraCharge: 0,
  handleAddRecipe: () => {},
  handleRemoveRecipe: () => {},
  headline: 'with Tomato & Lemon',
  id: '5f4d4a7e62fb0224951e7ec4',
  image:
    'https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_405,q_auto,w_720/hellofresh_s3/image/5f4d4a7e62fb0224951e7ec4-2fe03fc2.jpg',
  maxRecipesSelected :false,
  minRecipesSelected : true,
  name: 'Chicken Sausage & Spinach Ravioli',
  slug: 'chicken-sausage-spinach-ravioli',
  selected: 1,
  selectionLimit: 1,
  yields: 2,
};

describe('Recipe Card Component', () => {
  test('render recipe card', () => {
    render(<RecipeCard {...recipeCardProps} />);
    expect(screen.getByTestId('recipe-card')).toBeDefined();
  })
  test('render recipe name', () => {
    render(<RecipeCard {...recipeCardProps} />);
    expect(screen.getByTestId('recipe-card-name')).toBeDefined();
  })
  test('render recipe headline', () => {
    render(<RecipeCard {...recipeCardProps} />);
    expect(screen.getByTestId('recipe-card-headline')).toBeDefined();
  })
  test('render unselected recipe footer', async () => {
      const unSelectedProps = {...recipeCardProps};
      unSelectedProps.selected = 0;
      render(<RecipeCard props= {unSelectedProps} />);
      await screen.findByTestId('recipe-unselected')
  })
  test('render selected recipe footer', async () => {
    render(<RecipeCard {...recipeCardProps} />);
    await screen.findByTestId('recipe-selected')
  })
});

const unSelectedProps = {
  price : 100,
  recipeId : '5f4d4a7e62fb0224951e7ec4',
  isExtra: false,
  isDisabled : false,
  handleAddRecipe : () => {},
}

describe('Unselected Recipe Footer', () => {
  test('render button with text - Add', () => {
    render(<UnselectedRecipeFooter {...unSelectedProps} />);
    expect(screen.getByRole('button', { exact : true, name: 'Add'}))
  })
  test('render button with text - Add extra meal', () => {
    const minRecProps = {...unSelectedProps};
    minRecProps.isExtra = true
    render(<UnselectedRecipeFooter {...minRecProps} />);
    expect(screen.getByRole('button',{name: 'Add extra meal', exact: true}))
  })
  test('render Add button in disabled state', () => {
    const disabledProps = {...unSelectedProps};
    disabledProps.isDisabled = true
    render(<UnselectedRecipeFooter {...disabledProps} />);
    expect(screen.getByText(/Add/i).closest('button')).toBeDisabled();
  })
  test('render Price only if extraCharge is not 0', () => {
    const extraChargeProps = {...unSelectedProps};
    extraChargeProps.price = 0;
    render(<UnselectedRecipeFooter {...extraChargeProps} />);
    expect(screen.queryByTestId('recipe-special-price')).not.toBeInTheDocument();
  })
  test('render selected recipe footer on click of add button', () => {
    const funcProps = {...unSelectedProps};
    funcProps.handleAddRecipe = jest.fn();
    render(<UnselectedRecipeFooter {...funcProps} />);
    fireEvent.click(screen.getByRole('button',{name: /Add/i}));
    expect(funcProps.handleAddRecipe).toHaveBeenCalledWith(funcProps.recipeId, 1);
  })
  test('it should match Unselected Recipe Footer snapshot', ()=> {
    expect(render(<UnselectedRecipeFooter {...unSelectedProps} />)).toMatchSnapshot();
  });
});

const selectedProps = {
  recipeId : '5f4d4a7e62fb0224951e7ec4',
  selected : 1,
  yields : 2,
  isDisabled: false,
  handleAddRecipe : () => {},
  handleRemoveRecipe : () => {},
};
describe('Selected Recipe Footer', () => {

  test('render total selected in recipe', () => {
    render(<SelectedRecipeFooter {...selectedProps}/>);
    expect(screen.getByTestId('recipe-selected-desc')).toHaveTextContent(
      '1 in your box'
    )
  });
  test('render total servings in recipe', () => {
    render(<SelectedRecipeFooter {...selectedProps}/>);
    expect(screen.getByTestId('recipe-selected-servings')).toHaveTextContent(
      '(2 servings)'
    )
  });
  test('click on + should trigger a function', () => {
    const addProps = {...selectedProps};
    addProps.handleAddRecipe = jest.fn();
    render(<SelectedRecipeFooter {...addProps}/>);
    fireEvent.click(screen.getByRole('button', {name: 'increase-quantity', exact: true}));
    expect(addProps.handleAddRecipe).toHaveBeenCalledTimes(1);
  });
  test('click on - should trigger a function', () => {
    const removeProps = {...selectedProps};
    removeProps.handleRemoveRecipe = jest.fn();
    render(<SelectedRecipeFooter {...removeProps}/>);
    fireEvent.click(screen.getByRole('button', {name: 'decrease-quantity', exact: true}));
    expect(removeProps.handleRemoveRecipe).toHaveBeenCalledTimes(1);
  });

  test('it should match Selected Recipe Footer snapshot', ()=> {
    expect(render(<SelectedRecipeFooter {...selectedProps} />)).toMatchSnapshot();
  });
});
