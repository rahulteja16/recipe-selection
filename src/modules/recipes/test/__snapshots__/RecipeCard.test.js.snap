// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected Recipe Footer it should match Selected Recipe Footer snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnxRM sc-dlnjwi cFwRot"
        data-testid="recipe-selected"
        display="flex"
      >
        <button
          aria-label="decrease-quantity"
          class="sc-jSFjdj cBxnSP"
          title="Decrease quantity"
        >
          <svg
            class="sc-hKFxyN bKWFni"
            height="24"
            viewBox="0 0 32 32"
            width="24"
          >
            <path
              d="M16 3c7.18 0 13 5.82 13 13s-5.82 13-13 13S3 23.18 3 16 8.82 3 16 3zm0 1.5C9.649 4.5 4.5 9.649 4.5 16S9.649 27.5 16 27.5 27.5 22.351 27.5 16 22.351 4.5 16 4.5zM22 15v2H10v-2h12z"
            />
          </svg>
        </button>
        <div
          class="sc-bdnxRM cEwTyI"
          color="white"
        >
          <div
            class="sc-eCApnc gJzqRs"
            data-testid="recipe-selected-desc"
            font-family="secondary"
            font-size="md"
            font-weight="bold"
          >
            1
             in your box
          </div>
          <div
            class="sc-eCApnc eiZrLV"
            data-testid="recipe-selected-servings"
            font-family="secondary"
            font-size="sm"
          >
            (
            2
             servings)
          </div>
        </div>
        <button
          aria-label="increase-quantity"
          class="sc-jSFjdj cBxnSP"
          title="Increase quantity"
        >
          <svg
            class="sc-hKFxyN bKWFni"
            height="24"
            viewBox="0 0 32 32"
            width="24"
          >
            <path
              d="M16 3c7.18 0 13 5.82 13 13s-5.82 13-13 13S3 23.18 3 16 8.82 3 16 3zm0 1.5C9.649 4.5 4.5 9.649 4.5 16S9.649 27.5 16 27.5 27.5 22.351 27.5 16 22.351 4.5 16 4.5zm1 5.5v5h5v2h-5v5h-2v-5h-5v-2h5v-5h2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM sc-dlnjwi cFwRot"
      data-testid="recipe-selected"
      display="flex"
    >
      <button
        aria-label="decrease-quantity"
        class="sc-jSFjdj cBxnSP"
        title="Decrease quantity"
      >
        <svg
          class="sc-hKFxyN bKWFni"
          height="24"
          viewBox="0 0 32 32"
          width="24"
        >
          <path
            d="M16 3c7.18 0 13 5.82 13 13s-5.82 13-13 13S3 23.18 3 16 8.82 3 16 3zm0 1.5C9.649 4.5 4.5 9.649 4.5 16S9.649 27.5 16 27.5 27.5 22.351 27.5 16 22.351 4.5 16 4.5zM22 15v2H10v-2h12z"
          />
        </svg>
      </button>
      <div
        class="sc-bdnxRM cEwTyI"
        color="white"
      >
        <div
          class="sc-eCApnc gJzqRs"
          data-testid="recipe-selected-desc"
          font-family="secondary"
          font-size="md"
          font-weight="bold"
        >
          1
           in your box
        </div>
        <div
          class="sc-eCApnc eiZrLV"
          data-testid="recipe-selected-servings"
          font-family="secondary"
          font-size="sm"
        >
          (
          2
           servings)
        </div>
      </div>
      <button
        aria-label="increase-quantity"
        class="sc-jSFjdj cBxnSP"
        title="Increase quantity"
      >
        <svg
          class="sc-hKFxyN bKWFni"
          height="24"
          viewBox="0 0 32 32"
          width="24"
        >
          <path
            d="M16 3c7.18 0 13 5.82 13 13s-5.82 13-13 13S3 23.18 3 16 8.82 3 16 3zm0 1.5C9.649 4.5 4.5 9.649 4.5 16S9.649 27.5 16 27.5 27.5 22.351 27.5 16 22.351 4.5 16 4.5zm1 5.5v5h5v2h-5v5h-2v-5h-5v-2h5v-5h2z"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Unselected Recipe Footer it should match Unselected Recipe Footer snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnxRM sc-dlnjwi uFWrk"
        data-testid="recipe-unselected"
        display="flex"
      >
        <div
          class="sc-bdnxRM gbcpxT"
        >
          <div
            class="sc-eCApnc KdoYY"
            color="primary_600"
            data-testid="recipe-special-price"
          >
            +
            $1.00
          </div>
        </div>
        <div
          class="sc-bdnxRM iBwYVZ"
        >
          <button
            class="sc-gtsrHT bTcqaS"
            width="100%"
          >
            Add
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM sc-dlnjwi uFWrk"
      data-testid="recipe-unselected"
      display="flex"
    >
      <div
        class="sc-bdnxRM gbcpxT"
      >
        <div
          class="sc-eCApnc KdoYY"
          color="primary_600"
          data-testid="recipe-special-price"
        >
          +
          $1.00
        </div>
      </div>
      <div
        class="sc-bdnxRM iBwYVZ"
      >
        <button
          class="sc-gtsrHT bTcqaS"
          width="100%"
        >
          Add
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
